### PR TITLE
Improve startup exceptions (especially file permissions etc)

### PR DESCRIPTION
**Pull request by rmuir**
Its probably likely users will configure paths and have issues (file not found, bad file permissions, etc).
Today the failure is pitiful:

```
Exception in thread "main" /path
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:84)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
        at sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:384)
        at java.nio.file.Files.createDirectory(Files.java:674)
        at java.nio.file.Files.createAndCheckIsDirectory(Files.java:781)
        at java.nio.file.Files.createDirectories(Files.java:767)
        at org.elasticsearch.bootstrap.Security.ensureDirectoryExists(Security.java:171)
        at org.elasticsearch.bootstrap.Security.addPath(Security.java:151)
        at org.elasticsearch.bootstrap.Security.createPermissions(Security.java:133)
        at org.elasticsearch.bootstrap.Security.configure(Security.java:57)
        at org.elasticsearch.bootstrap.Bootstrap.setupSecurity(Bootstrap.java:188)
        at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:166)
        at org.elasticsearch.bootstrap.Bootstrap.doMain(Bootstrap.java:281)
        at org.elasticsearch.bootstrap.Bootstrap.main(Bootstrap.java:228)
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:28)
Refer to the log for complete error details.
```

We need more context here, like this:

```
Exception in thread "main" java.lang.IllegalStateException: Unable to access 'path.data' (/path/to/data)
Likely root cause: java.nio.file.AccessDeniedException: /path
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:84)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
        at sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:384)
        at java.nio.file.Files.createDirectory(Files.java:674)
        at java.nio.file.Files.createAndCheckIsDirectory(Files.java:781)
        at java.nio.file.Files.createDirectories(Files.java:767)
        at org.elasticsearch.bootstrap.Security.ensureDirectoryExists(Security.java:181)
        at org.elasticsearch.bootstrap.Security.addPath(Security.java:158)
        at org.elasticsearch.bootstrap.Security.createPermissions(Security.java:133)
        at org.elasticsearch.bootstrap.Security.configure(Security.java:57)
        at org.elasticsearch.bootstrap.Bootstrap.setupSecurity(Bootstrap.java:188)
        at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:166)
        at org.elasticsearch.bootstrap.Bootstrap.doMain(Bootstrap.java:281)
        at org.elasticsearch.bootstrap.Bootstrap.main(Bootstrap.java:228)
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:28)
Refer to the log for complete error details.
```

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/bootstrap/Security.java [hunk 1]
```diff
@@ -145,10 +145,20 @@ static Permissions createPermissions(Environment environment) throws IOException
         return policy;
     }
     
-    /** Add access to path (and all files underneath it */
-    static void addPath(Permissions policy, Path path, String permissions) throws IOException {
-        // paths may not exist yet
-        ensureDirectoryExists(path);
+    /**
+     * Add access to path (and all files underneath it
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/bootstrap/Security.java]: nit: missing ending paren

**Comment by dakrone** [comment]
LGTM

**Comment by rjernst** [comment]
LGTM too!

**Comment by rmuir** [comment]
Closed via https://github.com/elastic/elasticsearch/commit/d96af934db7a9ab25e8e76f01a2b5fe4c26f6ec1

**Pull request closed without merging**